src/system_identity/iam_client.rs

```diff
@@ -0,0 +1,189 @@
+use base64::{engine::general_purpose, Engine};
+use http::{
+    header::{AUTHORIZATION, CONTENT_TYPE},
+    StatusCode, Uri,
+};
+use serde::{Deserialize, Deserializer};
+use serde_json::json;
+use thiserror::Error;
+
+use crate::{
+    http_client::{HttpClient, HttpClientError},
+    token::AccessToken,
+};
+
+#[derive(Debug, Clone, Error)]
+pub enum IAMClientError {
+    #[error("error creating system identity: `{0}`")]
+    IAMClientError(String),
+}
+
+pub trait IAMClient {
+    fn create_system_identity(
+        &self,
+        token: &AccessToken,
+        pub_key: &[u8],
+    ) -> Result<SystemIdentityCreationResponseData, IAMClientError>;
+}
+
+/// Implementation of the IAMClient trait for a generic HTTP client.
+pub struct HttpIAMClientImpl<'a, C: HttpClient> {
+    http_client: &'a C,
+    name: String,
+    organization_id: String,
+    system_identity_creation_uri: &'a Uri,
+    // other required data? Like access token retriever or auth mechanism? key pair generator?
+}
+
+impl<'a, C: HttpClient> HttpIAMClientImpl<'a, C> {
+    pub fn new(
+        http_client: &'a C,
+        name: String,
+        organization_id: String,
+        system_identity_creation_uri: &'a Uri,
+    ) -> Self {
+        Self {
+            http_client,
+            name,
+            organization_id,
+            system_identity_creation_uri,
+        }
+    }
+
+    fn system_identity_from_token(
+        http_client: &C,
+        name: &str,
+        organization_id: &str,
+        token: &str,    // type
+        pub_key: &[u8], // type and provenance, other modes accepted?
+        system_identity_creation_uri: &Uri,
+    ) -> Result<SystemIdentityCreationResponseData, HttpClientError> {
+        let pub_key_b64 = general_purpose::STANDARD.encode(pub_key);
```

Comment:
I remember that the `pubkey` needed to be encoded in a particular way. Just for double-checking, have you tested if this work?

Reply:
I tested it by running a test with `reqwest::blocking::Client` and it worked, but I'll double-check. I might add some complete examples just as the ones you added.

Reply:
Just to clarify, I was meaning to actually create an identity in SIS ;-)